mlpp_lib/model_selection.py

```diff
@@ -1,10 +1,329 @@
+        station_split = {k: v["station"] for k, v in splits.items()}
+        splitter = cls(time_split, station_split)
+        splitter._time_defined = True
+        splitter._station_defined = True
+        splitter._time_partitioning()
+        splitter._station_partitioning()
+        return splitter
+
+    def to_json(self, file: str):
+        if not hasattr(self, "partitions"):
+            self._time_partitioning()
+            self._station_partitioning()
+        with open(file, "w") as f:
+            json.dump(self.partitions, f, indent=4)
+
+    def get_partition(
+        self, *args: xr.Dataset, partition=None, thinning: Optional[Mapping] = None
+    ) -> tuple[xr.Dataset, ...]:
+        """
+        Select and return a partition based on the current values in
+        the `partitions` attribute.
+
+        Parameters
+        ----------
+        *args: `xr.Dataset`
+            The dataset arguments from which we want to select a partition.
+        partition: str
+            The keyword for the partition, e.g. "train", "val" or "test".
+
+        Returns
+        -------
+        partition: tuple of `xr.Dataset`
+            Subsets of the input datasets.
+        """
+
+        if partition is None:
+            raise ValueError("Keyword argument `partition` must be provided.")
+
+        self.time_index = args[0].forecast_reference_time.values.copy()
+        self.station_index = args[0].station.values.copy()
+
+        self._time_partitioning()
+        self._station_partitioning()
+
+        # avoid out-of-order indexing (leads to bad performance with xarray/dask)
+        station_idx = self.partitions[partition]["station"]
+        idx_loc = [pd.Index(self.station_index).get_loc(label) for label in station_idx]
+        self.partitions[partition]["station"] = list(
+            np.array(station_idx)[np.argsort(idx_loc)]
+        )
+
+        # apply thinning
+        if thinning:
+            indexers = {
+                dim: coord[slice(None, None, thinning.get(dim, None))]
+                for dim, coord in self.partitions[partition].items()
+            }
+        else:
+            indexers = self.partitions[partition]
+
+        res = tuple(ds.sel(indexers) for ds in args)
+        for ds in args:
+            ds.close()
+        del args
+        return res
+
+    def _time_partitioning(self) -> None:
+
+        # actual computation of splits
+
+        if self._time_defined: # provided time split is all valid xarray indexers (lists or slices)
+            self._time_indexers = self.time_split
+        else:
+            self._time_indexers = {}
+            if all([isinstance(v, float) for v in self.time_split.values()]): # only fractions
+                res = self._time_partition_method(self.time_split)
+                self._time_indexers.update(res)
+            else: # mixed
+                _time_split = self.time_split.copy()
+                self._time_indexers.update({"test": _time_split.pop("test")})
```

Comment:
why hardcoding "test" here?

Reply:
This part of the code handles cases where the splits are provided as a mix of fractions for train and val datasets and slices/lists for test dataset (see docstrings). It's only executed if the test split exists and is already defined (i.e. not a fraction), so it's safe to have it hard coded